File: vito/core/src/main/java/com/facebook/fresco/vito/core/impl/CombinedImageListenerImpl.java
/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.facebook.fresco.vito.core.impl;

import android.graphics.drawable.Drawable;
import com.facebook.drawee.backends.pipeline.info.ImageOrigin;
import com.facebook.drawee.backends.pipeline.info.internal.ImagePerfControllerListener2;
import com.facebook.fresco.ui.common.BaseControllerListener2;
import com.facebook.fresco.ui.common.ControllerListener2;
import com.facebook.fresco.vito.core.CombinedImageListener;
import com.facebook.fresco.vito.core.VitoImageRequest;
import com.facebook.fresco.vito.core.VitoImageRequestListener;
import com.facebook.fresco.vito.core.VitoUtils;
import com.facebook.fresco.vito.listener.ImageListener;
import com.facebook.imagepipeline.image.ImageInfo;
import javax.annotation.Nullable;

public class CombinedImageListenerImpl implements CombinedImageListener {

  private @Nullable VitoImageRequestListener mVitoImageRequestListener;
  private @Nullable ImageListener mImageListener;
  private ControllerListener2<ImageInfo> mControllerListener2 =
      BaseControllerListener2.getNoOpListener();
  private @Nullable ImagePerfControllerListener2 mImagePerfControllerListener;

  @Override
  public void setImageListener(@Nullable ImageListener imageListener) {
    mImageListener = imageListener;
  }

  @Override
  public void setVitoImageRequestListener(
      @Nullable VitoImageRequestListener vitoImageRequestListener) {
    mVitoImageRequestListener = vitoImageRequestListener;
  }

  @Override
  @Nullable
  public ImageListener getImageListener() {
    return mImageListener;
  }

  @Override
  public void setControllerListener2(ControllerListener2<ImageInfo> controllerListener2) {
    mControllerListener2 = controllerListener2;
  }

  public void setImagePerfControllerListener(
      @Nullable ImagePerfControllerListener2 imagePerfControllerListener) {
    mImagePerfControllerListener = imagePerfControllerListener;
  }

  @Override
  public void onSubmit(
      long id,
      VitoImageRequest imageRequest,
      @Nullable Object callerContext,
      @Nullable ControllerListener2.Extras extras) {
    if (mVitoImageRequestListener != null) {
      mVitoImageRequestListener.onSubmit(id, imageRequest, callerContext, extras);
    }
    if (mImageListener != null) {
      mImageListener.onSubmit(id, callerContext);
    }
    String stringId = VitoUtils.getStringId(id);
    mControllerListener2.onSubmit(stringId, callerContext, extras);
    if (mImagePerfControllerListener != null) {
      mImagePerfControllerListener.onSubmit(stringId, callerContext, extras);
    }
  }

  @Override
  public void onPlaceholderSet(
      long id, VitoImageRequest imageRequest, @Nullable Drawable placeholder) {
    if (mVitoImageRequestListener != null) {
      mVitoImageRequestListener.onPlaceholderSet(id, imageRequest, placeholder);
    }
    if (mImageListener != null) {
      mImageListener.onPlaceholderSet(id, placeholder);
    }
  }

  @Override
  public void onFinalImageSet(
      long id,
      VitoImageRequest imageRequest,
      @ImageOrigin int imageOrigin,
      @Nullable ImageInfo imageInfo,
      @Nullable ControllerListener2.Extras extras,
      @Nullable Drawable drawable) {
    if (mVitoImageRequestListener != null) {
      mVitoImageRequestListener.onFinalImageSet(
          id, imageRequest, imageOrigin, imageInfo, extras, drawable);
    }
    if (mImageListener != null) {
      mImageListener.onFinalImageSet(id, imageOrigin, imageInfo, drawable);
    }
    String stringId = VitoUtils.getStringId(id);
    mControllerListener2.onFinalImageSet(stringId, imageInfo, extras);
    if (mImagePerfControllerListener != null) {
      mImagePerfControllerListener.onFinalImageSet(stringId, imageInfo, extras);
    }
  }

  @Override
  public void onIntermediateImageSet(
      long id, VitoImageRequest imageRequest, @Nullable ImageInfo imageInfo) {
    if (mVitoImageRequestListener != null) {
      mVitoImageRequestListener.onIntermediateImageSet(id, imageRequest, imageInfo);
    }
    if (mImageListener != null) {
      mImageListener.onIntermediateImageSet(id, imageInfo);
    }
    String stringId = VitoUtils.getStringId(id);
    mControllerListener2.onIntermediateImageSet(stringId, imageInfo);
    if (mImagePerfControllerListener != null) {
      mImagePerfControllerListener.onIntermediateImageSet(stringId, imageInfo);
    }
  }

  @Override
  public void onIntermediateImageFailed(
      long id, VitoImageRequest imageRequest, Throwable throwable) {
    if (mVitoImageRequestListener != null) {
      mVitoImageRequestListener.onIntermediateImageFailed(id, imageRequest, throwable);
    }
    if (mImageListener != null) {
      mImageListener.onIntermediateImageFailed(id, throwable);
    }
    String stringId = VitoUtils.getStringId(id);
    mControllerListener2.onIntermediateImageFailed(stringId);
    if (mImagePerfControllerListener != null) {
      mImagePerfControllerListener.onIntermediateImageFailed(stringId);
    }
  }

  @Override
  public void onFailure(
      long id,
      VitoImageRequest imageRequest,
      @Nullable Drawable error,
      @Nullable Throwable throwable,
      @Nullable ControllerListener2.Extras extras) {
    if (mVitoImageRequestListener != null) {
      mVitoImageRequestListener.onFailure(id, imageRequest, error, throwable, extras);
    }
    if (mImageListener != null) {
      mImageListener.onFailure(id, error, throwable);
    }
    String stringId = VitoUtils.getStringId(id);
    mControllerListener2.onFailure(stringId, throwable, extras);
    if (mImagePerfControllerListener != null) {
      mImagePerfControllerListener.onFailure(stringId, throwable, extras);
    }
  }

  @Override
  public void onRelease(long id, VitoImageRequest imageRequest, ControllerListener2.Extras extras) {
    if (mVitoImageRequestListener != null) {
      mVitoImageRequestListener.onRelease(id, imageRequest, extras);
    }
    if (mImageListener != null) {
      mImageListener.onRelease(id);
    }
    String stringId = VitoUtils.getStringId(id);
    mControllerListener2.onRelease(stringId, extras);
    if (mImagePerfControllerListener != null) {
      mImagePerfControllerListener.onRelease(stringId, extras);
    }
  }

  @Override
  public void onReset() {
    if (mImagePerfControllerListener != null) {
      mImagePerfControllerListener.resetState();
    }
  }
}

File: vito/core-common-impl/src/main/java/com/facebook/fresco/vito/core/impl/CombinedImageListenerImpl.kt
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.facebook.fresco.vito.core.impl

import android.graphics.drawable.Drawable
import com.facebook.drawee.backends.pipeline.info.ImageOrigin
import com.facebook.fresco.ui.common.BaseControllerListener2
import com.facebook.fresco.ui.common.ControllerListener2
import com.facebook.fresco.ui.common.ControllerListener2.Extras
import com.facebook.fresco.ui.common.ImagePerfNotifier
import com.facebook.fresco.ui.common.ImagePerfNotifierHolder
import com.facebook.fresco.ui.common.VitoUtils
import com.facebook.fresco.vito.core.CombinedImageListener
import com.facebook.fresco.vito.core.ImagePerfLoggingListener
import com.facebook.fresco.vito.core.VitoImageRequest
import com.facebook.fresco.vito.core.VitoImageRequestListener
import com.facebook.fresco.vito.listener.ImageListener
import com.facebook.imagepipeline.image.ImageInfo
import java.io.Closeable
import java.io.IOException

class CombinedImageListenerImpl : CombinedImageListener {

  private var localVitoImageRequestListener: VitoImageRequestListener? = null
  private var vitoImageRequestListener: VitoImageRequestListener? = null
  override var imageListener: ImageListener? = null
  private var controllerListener2: ControllerListener2<ImageInfo>? =
      BaseControllerListener2.getNoOpListener()
  private var imagePerfLoggingListener: ImagePerfLoggingListener? = null
  private var localImagePerfStateListener: ImagePerfNotifier? = null

  override fun setVitoImageRequestListener(vitoImageRequestListener: VitoImageRequestListener?) {
    this.vitoImageRequestListener = vitoImageRequestListener
  }

  override fun setLocalVitoImageRequestListener(
      vitoImageRequestListener: VitoImageRequestListener?
  ) {
    this.localVitoImageRequestListener = vitoImageRequestListener
  }

  override fun setControllerListener2(controllerListener2: ControllerListener2<ImageInfo>?) {
    this.controllerListener2 = controllerListener2
  }

  override fun setImagePerfLoggingListener(imagePerfLoggingListener: ImagePerfLoggingListener?) {
    this.imagePerfLoggingListener = imagePerfLoggingListener
    checkAndSetLocalImagePerfStateListener()
  }

  override fun getImagePerfLoggingListener(): ImagePerfLoggingListener? = imagePerfLoggingListener

  override fun setLocalImagePerfStateListener(imagePerfNotifier: ImagePerfNotifier?) {
    localImagePerfStateListener = imagePerfNotifier
    checkAndSetLocalImagePerfStateListener()
  }

  private fun checkAndSetLocalImagePerfStateListener() {
    val localPerfStatePublisher = imagePerfLoggingListener as? ImagePerfNotifierHolder
    if (localImagePerfStateListener != null && localPerfStatePublisher == null) {
      throw NullPointerException(
          "trying to set localImagePerfStateListener without a localPerfStatePublisher")
    }
    localPerfStatePublisher?.setImagePerfNotifier(localImagePerfStateListener)
  }

  override fun onSubmit(
      id: Long,
      imageRequest: VitoImageRequest,
      callerContext: Any?,
      extras: Extras?
  ) {
    vitoImageRequestListener?.onSubmit(id, imageRequest, callerContext, extras)
    localVitoImageRequestListener?.onSubmit(id, imageRequest, callerContext, extras)
    imageListener?.onSubmit(id, callerContext)
    val stringId = VitoUtils.getStringId(id)
    controllerListener2?.onSubmit(stringId, callerContext, extras)
    imagePerfLoggingListener?.onSubmit(stringId, callerContext, extras)
  }

  override fun onPlaceholderSet(id: Long, imageRequest: VitoImageRequest, placeholder: Drawable?) {
    vitoImageRequestListener?.onPlaceholderSet(id, imageRequest, placeholder)
    localVitoImageRequestListener?.onPlaceholderSet(id, imageRequest, placeholder)
    imageListener?.onPlaceholderSet(id, placeholder)
  }

  override fun onFinalImageSet(
      id: Long,
      imageRequest: VitoImageRequest,
      @ImageOrigin imageOrigin: Int,
      imageInfo: ImageInfo?,
      extras: Extras?,
      drawable: Drawable?
  ) {
    vitoImageRequestListener?.onFinalImageSet(
        id, imageRequest, imageOrigin, imageInfo, extras, drawable)
    localVitoImageRequestListener?.onFinalImageSet(
        id, imageRequest, imageOrigin, imageInfo, extras, drawable)
    imageListener?.onFinalImageSet(id, imageOrigin, imageInfo, drawable)
    val stringId = VitoUtils.getStringId(id)
    controllerListener2?.onFinalImageSet(stringId, imageInfo, extras)
    imagePerfLoggingListener?.onFinalImageSet(stringId, imageInfo, extras)
  }

  override fun onIntermediateImageSet(
      id: Long,
      imageRequest: VitoImageRequest,
      imageInfo: ImageInfo?
  ) {
    vitoImageRequestListener?.onIntermediateImageSet(id, imageRequest, imageInfo)
    localVitoImageRequestListener?.onIntermediateImageSet(id, imageRequest, imageInfo)
    imageListener?.onIntermediateImageSet(id, imageInfo)
    val stringId = VitoUtils.getStringId(id)
    controllerListener2?.onIntermediateImageSet(stringId, imageInfo)
    imagePerfLoggingListener?.onIntermediateImageSet(stringId, imageInfo)
  }

  override fun onIntermediateImageFailed(
      id: Long,
      imageRequest: VitoImageRequest,
      throwable: Throwable?
  ) {
    vitoImageRequestListener?.onIntermediateImageFailed(id, imageRequest, throwable)
    localVitoImageRequestListener?.onIntermediateImageFailed(id, imageRequest, throwable)
    imageListener?.onIntermediateImageFailed(id, throwable)
    val stringId = VitoUtils.getStringId(id)
    controllerListener2?.onIntermediateImageFailed(stringId)
    imagePerfLoggingListener?.onIntermediateImageFailed(stringId)
  }

  override fun onFailure(
      id: Long,
      imageRequest: VitoImageRequest,
      error: Drawable?,
      throwable: Throwable?,
      extras: Extras?
  ) {
    vitoImageRequestListener?.onFailure(id, imageRequest, error, throwable, extras)
    localVitoImageRequestListener?.onFailure(id, imageRequest, error, throwable, extras)
    imageListener?.onFailure(id, error, throwable)
    val stringId = VitoUtils.getStringId(id)
    controllerListener2?.onFailure(stringId, throwable, extras)
    imagePerfLoggingListener?.onFailure(stringId, throwable, extras)
  }

  override fun onRelease(id: Long, imageRequest: VitoImageRequest, extras: Extras?) {
    vitoImageRequestListener?.onRelease(id, imageRequest, extras)
    localVitoImageRequestListener?.onRelease(id, imageRequest, extras)
    imageListener?.onRelease(id)
    val stringId = VitoUtils.getStringId(id)
    controllerListener2?.onRelease(stringId, extras)
    imagePerfLoggingListener?.onRelease(stringId, extras)
  }

  override fun onEmptyEvent(callerContext: Any?) {
    vitoImageRequestListener?.onEmptyEvent(callerContext)
    localVitoImageRequestListener?.onEmptyEvent(callerContext)
    controllerListener2?.onEmptyEvent(callerContext)
    imagePerfLoggingListener?.onEmptyEvent(callerContext)
  }

  override fun onReset(
      resetVitoImageRequestListener: Boolean,
      resetLocalVitoImageRequestListener: Boolean,
      resetLocalImagePerfStateListener: Boolean,
      resetControllerListener2: Boolean,
  ) {
    try {
      imageListener = null
      (imagePerfLoggingListener as? Closeable)?.close()
      if (resetVitoImageRequestListener) {
        vitoImageRequestListener = null
      }
      if (resetLocalVitoImageRequestListener) {
        localVitoImageRequestListener = null
      }
      if (resetLocalImagePerfStateListener) {
        localImagePerfStateListener = null
      }
      if (resetControllerListener2) {
        controllerListener2 = null
      }
    } catch (e: IOException) {}
  }
}
